protocol_driver_double_barrel.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "protocol_driver_double_barrel.h"
+//#include "protocol_driver_allocator.h"
+#include "absl/base/internal/sysinfo.h"
+#include "distbench_utils.h"
+#include "glog/logging.h"
+#include "google/protobuf/repeated_field.h"
+
+namespace distbench {
+
+absl::StatusOr<std::unique_ptr<ProtocolDriver>> AllocateProtocolDriver(
+    ProtocolDriverOptions opts, int* port, int tree_depth = 0);
+
+// ProtocolDriver ===============================================
+ProtocolDriverDoubleBarrel::ProtocolDriverDoubleBarrel(int tree_depth) {
+  tree_depth_ = tree_depth;
+}
+
+absl::Status ProtocolDriverDoubleBarrel::Initialize(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+
+  auto pdo = pd_opts;
+  std::string subordinate_driver_type = "";
+  google::protobuf::internal::RepeatedPtrIterator
+      <distbench::NamedSetting> iter_subordinate;
+  for (auto iter=pdo.server_settings().begin();
+       iter!=pdo.server_settings().end();
+       iter++) {
+    if (iter->name()=="next_protocol_driver") {
+      subordinate_driver_type = iter->string_value();
+      break;
+    }
+  }
+  pdo.set_protocol_name(subordinate_driver_type);
+
+  auto maybe_barrel_1 = AllocateProtocolDriver(pdo, &port_1_, tree_depth_+1);
+  if (!maybe_barrel_1.ok()) return maybe_barrel_1.status();
+  barrel_1_ = std::move(maybe_barrel_1.value());
+
+  auto maybe_barrel_2 = AllocateProtocolDriver(pdo, &port_2_, tree_depth_+1);
+  if (!maybe_barrel_2.ok()) return maybe_barrel_2.status();
+  barrel_2_ = std::move(maybe_barrel_2.value());
+
+  return absl::OkStatus();
+}
+
+absl::Status ProtocolDriverDoubleBarrel::InitializeClient(
+    const ProtocolDriverOptions& pd_opts) {
+  return absl::UnimplementedError("InitializeClient is not implemented for double_barrel.");
```

Comment:
this should call the initialize method of the two subordinate protocol drivers, right?

Reply:
No, The Initialize for double barrel calls Initialize of the driver below it. It then calls InitializeClient and InitializeServer for actual driver

Reply:
If/when we spit the protocol drivers up into separate client and server classes this will be technical debt that has to be fixed later. Unless adding this code breaks things it's better to just do the simple thing and pass it through.